### PR TITLE
IODEMO: continue conn retries after responses timeout

### PR DESCRIPTION
**Pull request by evgeny-leksikov**

### CODE REVIEW

**Comment by evgeny-leksikov** [comment]
@yosefe could you pls review?

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -846,10 +856,12 @@ class DemoClient : public P2pDemoCommon {
 
             timersub(&tv_curr, &tv_start, &tv_diff);
             double elapsed = tv_diff.tv_sec + (tv_diff.tv_usec * 1e-6);
-            if (elapsed > _test_opts.client_timeout * 10) {
+            if ((elapsed > _test_opts.client_timeout * 10) && (_status = OK)) {
                 LOG << "timeout waiting for " << (_num_sent - _num_completed)
                     << " replies";
                 _status = WAIT_RESPONSES_TIMEOUT;
+                /* cancel all but wait counters */
+                do_io_cancel_all();
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: we should cancel (and close connection) only those servers which have uncompleted (e.g, timed-out) operations

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -918,19 +930,24 @@ class DemoClient : public P2pDemoCommon {
...
             // The active list is not empty, and we don't have to check the
             // connect retry timeout
+            _status = OK;
             return;
         }
 
         double curr_time = get_time();
         if (curr_time < (_prev_connect_time + opts().client_timeout)) {
             // Not enough time elapsed since previous connection attempt
+            if (_status == WAIT_RESPONSES_TIMEOUT) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe we can remove the status WAIT_RESPONSES_TIMEOUT altogether: if a connection times out, it triggers reconnect. if there are too many reconnects, it will fail with CONN_RETRIES_EXCEEDED

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -965,6 +982,11 @@ class DemoClient : public P2pDemoCommon {
             LOG << "Connected to " << server_name(server_index);
         }
 
+        /* all servers are connected or in progress, reset status to OK */
+        if (_active_servers.size() == _server_info.size()) {
+            _status = OK;
+        }
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: can we avoid this? set status to !=OK only when the detected error is really fatal

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -846,10 +854,10 @@ class DemoClient : public P2pDemoCommon {
 
             timersub(&tv_curr, &tv_start, &tv_diff);
             double elapsed = tv_diff.tv_sec + (tv_diff.tv_usec * 1e-6);
-            if (elapsed > _test_opts.client_timeout * 10) {
+            if (elapsed > _test_opts.client_timeout) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:894]: maybe increase the default timeout by 10x?

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -989,11 +998,7 @@ class DemoClient : public P2pDemoCommon {
             VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
 
             wait_for_responses(opts().window_size - 1);
-            if (_status != OK) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1023]: :+1:

File: test/apps/iodemo/io_demo.cc [hunk 6]
```diff
@@ -763,6 +762,14 @@ class DemoClient : public P2pDemoCommon {
         return data_size;
     }
 
+    void close_uncompleted_servers() {
+        for (size_t i = 0; i < _active_servers.size(); ++i) {
+            if (get_num_uncompleted(i) > 0) {
+                dispatch_connection_error(_server_info[i].conn);
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: let's have different log print in case of UCX-triggered error and IOdemo-triggered timeout

File: test/apps/iodemo/ucx_wrapper.h [hunk 7]
```diff
@@ -86,15 +86,13 @@ class UcxContext {
...
-
     // Called when new IO message is received
     virtual void dispatch_io_message(UcxConnection* conn, const void *buffer,
-                                     size_t length);
+                                     size_t length) = 0;
 
     // Called when there is a fatal failure on the connection
-    virtual void dispatch_connection_error(UcxConnection* conn);
+    virtual void dispatch_connection_error(UcxConnection* conn,
+                                           const char *reason) = 0;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.h]: do not modify ucx_wrapper
have 1 more function in DemoClient only which is called both from dispatch_connecton_error, and laso from timeout flow, which gets the reason as parameter

Comment (evgeny-leksikov) [reply test/apps/iodemo/ucx_wrapper.h]: I'm not sure that completely understand you, `dispatch_connection_error` is also called from `UcxContext::progress_failed_connections` so we have to know reason here too.

Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.h]:
```
class IOdemoCommon: public UcxContext
protected:
/// @override
virtual void dispatch_connection_error(UcxConnection *conn) {
  terminate_connection(conn, ucs_status_string(conn->get_status()));
}

void terminate_connection(conn, char* reason) {
 LOG << ... << reason;
 .. all cleanup code
}
}

class P2cclient: public IOdemoCommon {
void close_uncompleted_servers(const char *reason) {
 for (...) {
  if (get_num_uncompleted(i) > 0) {
   terminate_connection(server_info[i].conn, reason)
 }
}
}

wait_for_completions() {
...
}

```

File: test/apps/iodemo/io_demo.cc [hunk 8]
```diff
@@ -989,19 +1003,16 @@ class DemoClient : public P2pDemoCommon {
...
-            }
-
-            connect_all((total_iter % 10) == 0);
+            connect_all(is_control_iter(total_iter));
             if (_status != OK) {
                 break;
             }
 
             if (_active_servers.empty()) {
+                double retry_iterval = opts().client_timeout / 10;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe add new param: reconnect interval (could be in another pr)
BTW connect_all still uses client_timeout at line 956

File: test/apps/iodemo/io_demo.cc [hunk 9]
```diff
@@ -1098,6 +1107,10 @@ class DemoClient : public P2pDemoCommon {
                                  size_t(0), opts().operations.size() - 1)];
     }
 
+    static bool is_control_iter(long iter) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe move it before it's being used, and make make it inline?

File: test/apps/iodemo/ucx_wrapper.cc [hunk 10]
```diff
@@ -163,22 +163,6 @@ void UcxContext::progress()
...
-                                     size_t length)
-{
-    // To be implemented in a subclass
-}
-
-void UcxContext::dispatch_connection_error(UcxConnection* conn)
-{
-    // To be implemented in a subclass
-}
-
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:188]: really need to remove?

File: test/apps/iodemo/ucx_wrapper.cc [hunk 11]
```diff
@@ -265,7 +249,6 @@ void UcxContext::progress_conn_requests()
         UcxConnection *conn = new UcxConnection(*this, get_next_conn_id());
         if (conn->accept(_conn_requests.front())) {
             add_connection(conn);
-            dispatch_new_connection(conn);
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:275]: why removed?

File: test/apps/iodemo/ucx_wrapper.cc (hunk 10 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/ucx_wrapper.cc:188]: - easier to refactor maintain and debug
- actually dead code

File: test/apps/iodemo/ucx_wrapper.cc (hunk 11 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/ucx_wrapper.cc:275]: ~~dead code~~ noop

Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc:275]: ok

File: test/apps/iodemo/ucx_wrapper.cc (hunk 10 above)
Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc:188]: ok

**Review by yosefe — APPROVED** [review]
pls squash and rebase
